### PR TITLE
Add path details to headers on gql requests

### PR DESCRIPTION
**Pull request by gigxz**
## Description

Add path details to headers on gql requests

These headers are meant to support ~~https://github.com/greenriver/hmis-warehouse/pull/3700~~ https://github.com/greenriver/hmis-warehouse/pull/3720

To test with encoding, set `PUBLIC_PROTECTED_IDS='true'` in your `.env.development`

## Type of change
[//]: # 'remove options that are not relevant'
- [x] New feature (adds functionality)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (eslint)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: src/providers/apolloClient.tsx [hunk 1]
```diff
@@ -52,6 +56,37 @@ const authLink = setContext(
...
+
+    // Add header for applicable params, if present
+    const { clientId, enrollmentId, projectId } = decodedParams;
+    if (clientId) modifiedHeaders['X-Hmis-Client-Id'] = clientId;
+    if (enrollmentId) modifiedHeaders['X-Hmis-Enrollment-Id'] = enrollmentId;
+    if (projectId) modifiedHeaders['X-Hmis-Project-Id'] = projectId;
+    return { headers: modifiedHeaders };
+  }
+);
+
```
Comment (gigxz) [inline src/providers/apolloClient.tsx]: This is the important part

File: src/routes/routes.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
+import { flatMap } from 'lodash-es';
```
Comment (ttoomey) [inline src/routes/routes.ts]: I don't think we need lodash for this anymore
https://caniuse.com/?search=flatMap

File: src/utils/pathEncoding.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import { NotFoundError } from '@/modules/errors/util';
+import IdEncoder from '@/modules/hmis/IdEncoder';
+import { HmisEnums } from '@/types/gqlEnums';
+
+export const isIdParam = (key: string) => key.match(/Id$/);
+
+const enumParams: Record<string, string[]> = {
+  formRole: Object.keys(HmisEnums.FormRole),
+};
+
```
Comment (ttoomey) [inline src/utils/pathEncoding.tsx:12]: maybe some docs here would be helpful
```suggestion
/**
 * Decodes encrypted ID parameters
 * Converts enum parameter values to uppercase and validates them
 * Throws a NotFoundError for unrecognized enum parameter values
 */
```

File: src/providers/apolloClient.tsx [hunk 4]
```diff
@@ -52,6 +56,37 @@ const authLink = setContext(
...
+  (
+    _,
+    { headers }: { headers: { [key: string]: string } }
+  ): { headers: Record<string, string> } => {
+    const modifiedHeaders: typeof headers = {
+      ...headers,
+      'X-Hmis-Path': window.location.pathname,
+    };
+
+    // Find whith Route matches the current location
```
Comment (ttoomey) [inline src/providers/apolloClient.tsx]:
```suggestion
    // Find when Route matches the current location
```

**Review by ttoomey — APPROVED** [review]
Looks great!

File: src/routes/routes.ts (hunk 2 above)
Comment (gigxz) [reply src/routes/routes.ts]: thanks! fixed

**Pull request merged**